demonstrations/tutorial_quantum_volume.py

```diff
@@ -0,0 +1,732 @@
+# IBM's 27-qubit superconducting device `ibmq_montreal` [#qv64]_, and
+# :math:`\log_2 V_Q = 7` on a Honeywell trapped-ion qubit processor
+# [#honeywell]_. A device with an expected quantum volume of :math:`\log_2 V_Q
+# = 32` has also been announced by IonQ [#ionq]_, though benchmarking results
+# have not yet been published.
+#
+# .. note::
+#
+#    In many sources, the quantum volume of processors is reported as
+#    :math:`V_Q` explicitly, rather than :math:`\log_2 V_Q` as is the
+#    convention in this demo. As such, IonQs processor has the potential for a
+#    quantum volume of :math:`2^{32} > 4000000`. Here we use the :math:`\log`
+#    because it is more straightforward to understand that they have 32
+#    high-quality, well-connected qubits than to extract this at first glance from the
+#    explicit value of the volume.
+#
+
+
+##############################################################################
+#
+# Computing the quantum volume
+# ----------------------------
+#
+# Equipped with our definition of quantum volume, it's time to compute it
+# ourselves!. We'll use the `PennyLane-Qiskit
+# <https://pennylaneqiskit.readthedocs.io/en/latest/>`_ plugin to compute the
+# volume of one of the IBM processors, since their properties are easily
+# accessible through this interface.
+#
+#
+# Loosely, the protocol for quantum volume consists of three steps:
+#
+# 1. Construct random square circuits of increasing size
+#
+# 2. Run those circuits on both a simulator, and on a noisy hardware device
+#
+# 3. Perform a statistical analysis of the results to determine what size
+#    circuits the device can run reliably
+#
+#
+# The largest reliable size will become the :math:`m` in the expression for
+# quantum volume.
+#
+#
+# Step 1: construct random square circuits
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Recall that the structure of the circuits above is alternating layers of
+# permutations and random SU(4) operations on pairs of qubits.  Let's implement
+# the generation of such circuits in PennyLane.
+#
+# First we write a function that randomly permutes qubits. We'll do this by
+# using numpy to generate a permutation, and then apply it with SWAP gates.
+
+import numpy as np
+import pennylane as qml
+
+# Object for random number generation from numpy
+np.random.seed(42)
+rng = np.random.default_rng()
+
+def permute_qubits(num_qubits):
+    # A random permutation
+    perm_order = list(rng.permutation(num_qubits))
+
+    working_order = list(range(num_qubits))
+
+    # We will permute by iterating through the permutation and swapping
+    # things back to their proper place.
+    for idx_here in range(num_qubits):
+        if working_order[idx_here] != perm_order[idx_here]:
+            # Where do we need to send the qubit at this location?
+            idx_there = working_order.index(perm_order[idx_here])
+            qml.SWAP(wires=[idx_here, idx_there])
```

Comment:
```suggestion
            idx_there = working_order.index(perm_order[idx_here])

            qml.SWAP(wires=[idx_here, idx_there])
```

(only because it took me a while to spot the swap, I couldn't work out what the function was returning!